core/src/banking_stage.rs

```diff
@@ -543,154 +527,102 @@ impl BankingStage {
-                            if let Some(bank) = &end_of_slot.working_bank {
-                                let new_unprocessed_indexes =
-                                    Self::filter_unprocessed_packets_at_end_of_slot(
-                                        bank,
-                                        packet_batch,
-                                        original_unprocessed_indexes,
-                                        my_pubkey,
-                                        end_of_slot.next_slot_leader,
-                                        banking_stage_stats,
-                                    );
-
-                                let end_of_slot_filtered_invalid_count =
-                                    original_unprocessed_indexes
-                                        .len()
-                                        .saturating_sub(new_unprocessed_indexes.len());
-
-                                slot_metrics_tracker.increment_end_of_slot_filtered_invalid_count(
-                                    end_of_slot_filtered_invalid_count as u64,
-                                );
-
-                                banking_stage_stats
-                                    .end_of_slot_filtered_invalid_count
-                                    .fetch_add(
-                                        end_of_slot_filtered_invalid_count,
-                                        Ordering::Relaxed,
-                                    );
-
-                                Self::update_buffered_packets_with_new_unprocessed(
-                                    original_unprocessed_indexes,
-                                    new_unprocessed_indexes,
-                                )
-                            } else {
-                                true
-                            }
-                        },
-                        (),
-                        "end_of_slot_filtering",
+        RetainMut::retain_mut(buffered_packet_batches, |deserialized_packet_batch| {
+            let packet_batch = &deserialized_packet_batch.packet_batch;
+            let original_unprocessed_indexes = deserialized_packet_batch
+                .unprocessed_packets
+                .keys()
+                .cloned()
+                .collect::<Vec<usize>>();
+            if let Some(end_of_slot) = &reached_end_of_slot {
+                // We've hit the end of this slot, no need to perform more processing,
+                // just filter the remaining packets for the invalid (e.g. too old) ones
+                // if the working_bank is available
+                let mut end_of_slot_filtering_time = Measure::start("end_of_slot_filtering");
+                let should_retain = if let Some(bank) = &end_of_slot.working_bank {
+                    let new_unprocessed_indexes = Self::filter_unprocessed_packets_at_end_of_slot(
+                        bank,
+                        packet_batch,
```

Comment:
can we pass the `deserialize_packet_batch.unprocessed_packets` here and rework process_packets_transactions to use those deserialized packets instead so we don't have to incur the cost of deserializing these packets again?